datastores/gossip_kv/protocol/lib.rs

```diff
@@ -0,0 +1,149 @@
+use std::fmt::Display;
+use std::str::FromStr;
+
+use serde::{Deserialize, Serialize};
+
+use crate::KeyParseError::InvalidNamespace;
+
+/// The namespace of the key of an entry in the key-value store.
+#[derive(Debug, Eq, PartialEq, Clone, Copy, Serialize, Deserialize, Hash)]
+pub enum Namespace {
+    /// User namespace is for use by the user of the key-value store.
+    User,
+
+    /// System namespace is reserved for use by the key-value store itself.
+    System,
+}
+
+/// Error that can occur when parsing a key from a string.
+#[derive(Debug, Eq, PartialEq, Clone, Copy)]
+pub enum KeyParseError {
+    /// The namespace in the key is invalid. Namespaces must be either `usr` or `sys`.
+    InvalidNamespace,
+
+    /// The key is in an invalid format. Keys must be of the form `/namespace/table/row`.
+    InvalidFormat,
+}
+
+impl Display for KeyParseError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            InvalidNamespace => write!(f, "Invalid namespace"),
+            KeyParseError::InvalidFormat => write!(f, "Invalid key format"),
+        }
+    }
+}
+
+impl FromStr for Namespace {
+    type Err = KeyParseError;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "usr" => Ok(Namespace::User),
+            "sys" => Ok(Namespace::System),
+            _ => Err(InvalidNamespace),
+        }
+    }
+}
+
+/// The name of a table in the key-value store.
+pub type TableName = String;
+
+/// The key of a row in a table in the key-value store.
+pub type RowKey = String;
+
+/// A key of an entry in the key-value store.
+///
+/// Data in the key-value store is organized into namespaces, tables, and rows. Namespaces are
+/// either `usr` for user data or `sys` for system data. Namespaces contain tables, which contain
+/// rows. Each row has a row key and a row value.
+#[derive(Debug, Eq, PartialEq, Clone, Serialize, Deserialize, Hash)]
+pub struct Key {
+    /// The namespace of the key.
+    pub namespace: Namespace,
+    /// The name of the table in the key.
+    pub table: TableName,
+    /// The key of the row in the table.
+    pub row_key: RowKey,
+}
+
+impl FromStr for Key {
+    type Err = KeyParseError;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        // TODO: Support escaping '/' in table and row keys. https://github.com/hydro-project/hydroflow/issues/1254
+        let parts: Vec<&str> = s.split('/').collect();
+        if parts.len() != 4 {
+            return Err(KeyParseError::InvalidFormat);
+        }
+        let namespace = parts[1].parse()?;
+        Ok(Key {
+            namespace,
+            table: parts[2].to_string(),
+            row_key: parts[3].to_string(),
+        })
```

Comment:
What is `parts[0]`? Should ensure it is empty string?

Reply:
Good point.